File: ts-contacts-service/src/main/java/contacts/service/FeatureFlagService.java
package contacts.service;

import dev.openfeature.sdk.Client;
import dev.openfeature.sdk.FlagEvaluationDetails;
import dev.openfeature.sdk.OpenFeatureAPI;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;

@Service
public class FeatureFlagService {
    
    private Client client;
    private boolean isInitialized = false;
    
    @PostConstruct
    public void initialize() {
        try {
            // Get a named client for contacts-service
            this.client = OpenFeatureAPI.getInstance().getClient("contacts-service");
            System.out.println("[TrainTicket][Contacts][FeatureFlagService] Initialized successfully");
            this.isInitialized = true;
        } catch (Exception e) {
            System.err.println("[TrainTicket][Contacts][FeatureFlagService] Failed to initialize: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    public boolean isEnabled(String flagName) {
        // Simple fix: If not initialized yet, return false (safe default)
        if (!this.isInitialized || this.client == null) {
            System.out.println("[TrainTicket][Contacts][FeatureFlagService] Client not ready yet, using safe default (false) for flag: " + flagName);
            return false;
        }
        
        try {
            FlagEvaluationDetails<Boolean> details = client.getBooleanDetails(flagName, false);
            System.out.println(String.format(
                "[TrainTicket][Contacts][FeatureFlagService] Flag %s: value=%s, reason=%s", 
                flagName, 
                details.getValue(), 
                details.getReason() != null ? details.getReason() : "N/A"
            ));
            
            // Check for ERROR reason
            if ("ERROR".equals(details.getReason())) {
                System.err.println("[TrainTicket][Contacts][FeatureFlagService] Provider error for flag " + flagName);
                return false;
            }
            
            return Boolean.TRUE.equals(details.getValue());
            
        } catch (Exception e) {
            System.err.println("[TrainTicket][Contacts][FeatureFlagService] Error getting flag " + flagName + ": " + e.getMessage());
            return false;
        }
    }
}

File: ts-voucher-service/feature_flag_service.py
import os
from openfeature import api
from openfeature.contrib.provider.flagd import FlagdProvider
from openfeature.contrib.provider.flagd.config import ResolverType, CacheType


class FeatureFlagService:
    def __init__(self) -> None:
        provider = FlagdProvider(
            resolver_type=ResolverType.RPC,
            host="flagd",
            port=8013,
            cache=CacheType.DISABLED,
            max_cache_size=0,
        )
        api.set_provider(provider)
        self.client = api.get_client("voucher-service")

    def is_enabled(self, flag_name: str) -> bool:
        try:
            details = self.client.get_boolean_details(flag_name, False)
            print(
                f"[TrainTicket][Voucher][FeatureFlagService] Flag {flag_name}: value={details.value}, reason={getattr(details, 'reason', 'N/A')}"
            )

            if getattr(details, "reason", None) == "ERROR":
                print(
                    f"[TrainTicket][Voucher][FeatureFlagService] Provider error for flag {flag_name}"
                )
                return False
            return bool(details.value)
        except Exception as e:
            print(
                f"[TrainTicket][Voucher][FeatureFlagService] Error getting flag {flag_name}: {e}"
            )
            return False
